File: 트리/leetcode226(Easy)-이진 트리 반전.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

// 재귀 : 72ms(35.73%), 42.2MB(58.73%)
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var invertTree = function (root) {
  if (!root) {
    return null
  }

  const left = invertTree(root.right)
  const right = invertTree(root.left)
  root.left = left
  root.right = right
  return root
}

// 스택 : 68ms(53.21%), 42.4MB(35.99%)
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var invertTree = function (root) {
  const stack = [root]

  while (stack.length > 0) {
    const node = stack.pop()
    if (node) {
      const temp = node.right
      node.right = node.left
      node.left = temp
      stack.push(node.left)
      stack.push(node.right)
    }
  }

  return root
}

File: 트리/leetcode226(Easy)-이진 트리 반전.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# 재귀로 스왑 - 바텀업 방식 : 34ms(59.87%), 13.8MB(94.98%)
class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if root:
            root.left, root.right = self.invertTree(root.right), self.invertTree(root.left)
            return root
        return None

# 반복구조 BFS로 스왑 - 탑다운 방식 : 26ms(94.89%), 13.8MB(94.98%)
class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        q = collections.deque([root])
        while q:
            node = q.popleft()
            if node:
                node.left, node.right = node.right, node.left
                q.append(node.left)
                q.append(node.right)
        return root

# 반복구조 DFS로 스왑(전위순회) - 탑다운방식, 사실 큐로 푸나 스택으로 푸나 차이 없음 // 16ms(85%)
class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        stack = [root]
        while stack:
            node = stack.pop()
            if node:
                node.left, node.right = node.right, node.left
                stack.append(node.left)
                stack.append(node.right)
        return root

# 반복구조 DFS 후위순회
class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        stack = [root]
        while stack:
            node = stack.pop()
            if node:
                stack.append(node.left)
                stack.append(node.right)
                node.left, node.right = node.right, node.left
        return root